Changes made to runtime and environment of cluster tasks (slurm and sge)

FilomenoSanchez opened the pull request:
Just did some small changes, now the runtime can be set on slurm and the environment can be specified on SGE. Default value for the environment is "mpi".  I've tested this and it seems to work fine.

FilomenoSanchez commented:
Later commits focus on solving test failures

fsimkovic commented:
Please ensure all tests pass. How did you test this for Slurm/PBS and SGE? Did you try different versions? Do we need to restrict/advice on the tested versions?

fsimkovic commented on pyjob/sge.py:
```diff
@@ -104,13 +104,15 @@ def _create_runscript(self):
             runscript.append(self.__class__.SCRIPT_DIRECTIVE + ' ' + cmd)
         if self.runtime:
-            cmd = '-l h_rt={}'.format(self.runtime)
+            h, m = divmod(self.runtime, 60)
```
given that you code duplicated this from PBS, you should really refactor this section to a function that accepts the runtime and returns the string. there should be tests and then we can ensure this behaves correctly.

FilomenoSanchez replied:
I agree, I will work on this.

fsimkovic commented on pyjob/sge.py:
```diff
@@ -104,13 +104,15 @@ def _create_runscript(self):
             runscript.append(self.__class__.SCRIPT_DIRECTIVE + ' ' + cmd)
         if self.nprocesses:
-            cmd = '-pe mpi {}'.format(self.nprocesses)
+            cmd = '-pe {} {}'.format(self.environment, self.nprocesses)
```
what options are allowed? do we need to validate what we should allow for environment or trust the user to provide something reasonable

FilomenoSanchez replied:
The list of available environments depends on the specific configuration of the system, so it varies from cluster to cluster. A complete list of environments can be obtained using "qconf -spl". I guess it would be possible to create a function to run this command and only allow for environments listed with this command. This would actually solve a bug observed when running tasks through pyjob on a cluster with no specific environment setup (hpc02-ccp4 cluster, all cores allocated in the same node and no environment has been created). This cluster has no environments, and default "#$ -pe mpi 1" card crashes the task.

FilomenoSanchez commented:
> Please ensure all tests pass. How did you test this for Slurm/PBS and SGE? Did you try different versions? Do we need to restrict/advice on the tested versions?

I've got access to Slurm (version 17.02.8) and SGE (version 6.1.1) clusters, changes where tested on these. I do not have access to PBS clusters, but changes here only involve getting runtime into the hh:mm:ss format (02:00:00 instead of 2:0:0). This is also the reason why some tests failed, I can revert this modification if preferred, I just made this change because it seems more consistent.

FilomenoSanchez commented:
I've refactored the code as suggested so that the task's runtime is set using a `pyjob.task.Task` static method called `get_time`; also added new test for this method. Later commits focus on solving failed tests, had to make changes to accommodate new runtime format < hh:mm:ss > (asserting '02:00:00' instead of '2:0:0'). At the moment all tests pass except AppVeyor and Travis py2.7, both drop same error at `python setup.py test`. Not sure if this is related at all with the changes I've made, please check...

FilomenoSanchez commented:
All tests pass now. Also changed the `cexec` argument `directory` (it is deprecated) for `cwd`. This was causing issue #31 , which is now solved.
